hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -183,105 +270,166 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
+
+          if (compactMOBs) {
+            if (MobUtils.isMobReferenceCell(c)) {
+              String fName = MobUtils.getMobFileName(c);
+              Path pp = new Path(new Path(fs.getUri()), new Path(path, fName));
+
+              // Added to support migration
+              try {
+                mobCell = mobStore.resolve(c, true, false);
+              } catch (FileNotFoundException fnfe) {
+                if (discardMobMiss) {
+                  LOG.error("Missing MOB cell: file=" + pp + " not found");
+                  continue;
+                } else {
+                  throw fnfe;
+                }
+              }
+
+              if (discardMobMiss && mobCell.getCell().getValueLength() == 0) {
+                LOG.error("Missing MOB cell value: file=" + pp + " cell=" + mobCell);
+                continue;
+              }
+
+              if (mobCell.getCell().getValueLength() > mobSizeThreshold) {
+                // put the mob data back to the store file
+                PrivateCellUtil.setSequenceId(mobCell.getCell(), c.getSequenceId());
+                if (generationalMob) {
+                  //TODO: verify fName
+                  StoreFileWriter stw = mobWriters.getOutputWriterForInputFile(fName);
+                  if (stw != null) {
+                    stw.append(mobCell.getCell());
+                    mobWriters.incrementMobCountForOutputWriter(stw, 1);
+                  } else {
+                    // leave mob cell in a MOB file which is not in compaction selections
+                  }
+                } else {
+                  mobFileWriter.append(mobCell.getCell());
+                  mobCells++;
+                }
+                writer.append(MobUtils.createMobRefCell(mobCell.getCell(), fileName,
+                  this.mobStore.getRefCellTags()));
+                cellsCountCompactedFromMob++;
+                cellsSizeCompactedFromMob += mobCell.getCell().getValueLength();
+              } else {
+
+                // If MOB value is less than threshold, append it directly to a store file
+                PrivateCellUtil.setSequenceId(mobCell.getCell(), c.getSequenceId());
+                writer.append(mobCell.getCell());
+              }
+
             } else {
-              // Add a ref tag to this cell and write it to a store file.
-              writer.append(MobUtils.createMobRefDeleteMarker(c));
-              // Write the cell to a del file
-              delFileWriter.append(c);
-              deleteMarkersCount++;
+              // Not a MOB reference cell
+              int size = c.getValueLength();
+              if (size > mobSizeThreshold) {
+                // This MOB cell comes from a regular store file
+                // therefore we store it in original mob output
+                mobFileWriter.append(c);
+                writer
+                    .append(MobUtils.createMobRefCell(c, fileName, this.mobStore.getRefCellTags()));
+                mobCells++;
```

Comment:
this should increment the cells*CompactedToMob counters

Reply:
Fixed.